src/lxc/conf.c

```diff
@@ -1455,48 +1455,92 @@ static int setup_pivot_root(const struct lxc_rootfs *rootfs)
 
 static int setup_pts(int pts)
 {
+	bool created_ptmx = false;
+	int ret;
 	char target[PATH_MAX];
+	char *devpts_mntopts = "newinstance,ptmxmode=0666,mode=0620,gid=5";
 
-	if (!pts)
+	if (!pts) {
+		DEBUG("No new devpts instance will be mounted since no pts "
+		      "devices are requested.");
 		return 0;
-
-	if (!access("/dev/pts/ptmx", F_OK) && umount("/dev/pts")) {
-		SYSERROR("failed to umount 'dev/pts'");
-		return -1;
 	}
 
-	if (mkdir("/dev/pts", 0755)) {
-		if ( errno != EEXIST ) {
-		    SYSERROR("failed to create '/dev/pts'");
-		    return -1;
+	ret = access("/dev/pts/ptmx", F_OK);
+	if (!ret) {
+		/* Unmount old devpts instance. */
+		ret = umount("/dev/pts");
+		if (ret < 0) {
+			SYSERROR("Failed to unmount old devpts instance.");
+			return -1;
 		}
+		DEBUG("Unmounted old /dev/pts instance.");
 	}
 
-	if (mount("devpts", "/dev/pts", "devpts", MS_MGC_VAL,
-		  "newinstance,ptmxmode=0666,mode=0620,gid=5")) {
-		SYSERROR("failed to mount a new instance of '/dev/pts'");
+	/* Create mountpoint for devpts instance. */
+	ret = mkdir("/dev/pts", 0755);
+	if (ret < 0 && errno != EEXIST) {
+		SYSERROR("Failed to create the \"/dev/pts\" directory.");
 		return -1;
 	}
 
-	if (access("/dev/ptmx", F_OK)) {
-		if (!symlink("/dev/pts/ptmx", "/dev/ptmx"))
-			goto out;
-		SYSERROR("failed to symlink '/dev/pts/ptmx'->'/dev/ptmx'");
+	/* Mount new devpts instance. */
+	ret = mount("devpts", "/dev/pts", "devpts", MS_MGC_VAL, devpts_mntopts);
+	if (ret < 0) {
+		SYSERROR("Failed to mount new devpts instance.");
 		return -1;
 	}
 
-	if (realpath("/dev/ptmx", target) && !strcmp(target, "/dev/pts/ptmx"))
-		goto out;
+	ret = access("/dev/ptmx", F_OK);
+	if (ret < 0) {
+		/* Create dummy /dev/ptmx file as bind mount target for
+		 * /dev/pts/ptmx.
+		 */
+		ret = open("/dev/ptmx", O_CREAT, 0666);
+		if (ret < 0) {
+			SYSERROR("Failed to create dummy \"/dev/ptmx\" file as bind mount target.");
+			return -1;
+		}
+		created_ptmx = true;
+		DEBUG("Created dummy \"/dev/ptmx\" file as bind mount target.");
+	}
 
-	/* fallback here, /dev/pts/ptmx exists just mount bind */
-	if (mount("/dev/pts/ptmx", "/dev/ptmx", "none", MS_BIND, 0)) {
-		SYSERROR("mount failed '/dev/pts/ptmx'->'/dev/ptmx'");
+	/* Bind mount /dev/pts/ptmx to /dev/ptmx. */
+	ret = mount("/dev/pts/ptmx", "/dev/ptmx", "none", MS_BIND, 0);
```

Comment:
So what if /dev/ptmx exists as a symlink to /dev/pts/ptmx already?

The mount() call on it will deref the symlink and cause over-mounting of /dev/pts/ptmx won't it?